File: messaging-server-node/src/controllers/auth.controller.js
const { register, verify, validateLogin } = require('../services/user.service');
const { generateToken } = require('../helpers/jwt.helper');
const { otpCode } = require('../config');

const registerUser = (req, res) => {
  const { username, password } = req.body;
  if (register(username, password)) {
    return res.json({ message: 'Registered' });
  }
  return res.status(409).json({ error: 'User already exists' });
};

const verifyUser = (req, res) => {
  const { username, otp } = req.body;
  if (verify(username, otp, otpCode)) {
    return res.json({ message: 'Verified' });
  }
  return res.status(400).json({ error: 'Invalid OTP or user' });
};


const loginUser = (req, res) => {
  const { username, password } = req.body;
  if (!validateLogin(username, password))
    return res.status(401).send('Unauthorized');

  const token = generateToken(username);
  res.json({ "data" : token });
};

module.exports = { registerUser, verifyUser, loginUser };

File: MessagingServiceAppAPI/Controllers/AuthController.cs
using MessagingServiceAppAPI.Models;
using MessagingServiceAppAPI.Services;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace MessagingServiceAppAPI.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class AuthController : ControllerBase
    {
        private readonly AuthService _authService;
        private readonly IConfiguration _config;

        public AuthController(AuthService authService, IConfiguration config)
        {
            _authService = authService;
            _config = config;
        }


        [HttpPost("register")]
        public IActionResult Register(User user)
        {
            bool status = _authService.Register(user.Username, user.Password);
        return BuildResponse(
            status: status,
            data: null,
            message: status ? "Registered" : "User already exists");
        }

        [HttpPost("verify")]
        public IActionResult Verify(VerifyRequest data)
        {
            bool status = _authService.Verify((string)data.Username, (string)data.Otp);
        return BuildResponse(
            status: status,
            data: null,
            message: status ? "Verified" : "UnVerified");
        }

        [HttpPost("login")]
        public IActionResult Login(User user)
        {
            bool credentialsOk = _authService.ValidateLogin(user.Username,
                                                             user.Password);

            if (!credentialsOk)
                return BuildResponse(false, null, "Invalid username or password");

            string token = GenerateJwtToken(user.Username);
            return BuildResponse(true, new { token }, "Login successful");
        }

        private string GenerateJwtToken(string username)
        {
            var key = Encoding.UTF8.GetBytes(_config["Jwt:Key"]);
            var creds = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256);
            var token = new JwtSecurityToken(claims: new[] { new Claim(ClaimTypes.Name, username) }, expires: DateTime.UtcNow.AddHours(1), signingCredentials: creds);
            return new JwtSecurityTokenHandler().WriteToken(token);
        }

        private IActionResult BuildResponse(bool status, object? data, string message)
        {
            var payload = new
            {
                status,
                data,
                message
            };
            return Ok(payload);
        }
    }
}
